finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/StreamProxy.scala

```diff
@@ -1,23 +1,84 @@
- * The `onFrame` function is called for each frame in the stream.
- * @param underlying the [[Stream]] wrapped by this proxy
- * @param onFrame function called for each [[Frame]] in the underlying [[Stream]]
- */
-class StreamProxy(underlying: Stream, onFrame: Try[Frame] => Unit) extends Stream {
+abstract class StreamProxy(underlying: Stream) extends Stream {
+  /**
+   * @return true if the underlying stream is empty
+   */
   override def isEmpty: Boolean = underlying.isEmpty
-
-  override def read(): Future[Frame] = underlying.read().respond(onFrame)
-
   /**
    * Satisfied when an end-of-stream frame has been read from this
    * stream.
    *
    * If the stream is reset prematurely, onEnd fails with a [[Reset]].
    */
   override def onEnd: Future[Unit] = underlying.onEnd
+}
+/**
+ * Wraps an underlying [[Stream]] with an [[StreamOnFrame.onFrame onFrame]]
+ * function.
+ * The `onFrame` function is called for each frame in the stream.
+ *
+ * @param underlying the [[Stream]] wrapped by this proxy.
+ * @param onFrame function called for each [[Frame]] in the underlying [[Stream]].
+ */
+// TODO: consider renaming `onFrame` to `foreach`?
+class StreamOnFrame(underlying: Stream, onFrame: Try[Frame] => Unit)
+  extends StreamProxy(underlying) {
+  override def read(): Future[Frame] = underlying.read().respond(onFrame)
   override def toString: String = s"StreamProxy($underlying, onFrame=$onFrame)"
 }
+
+/**
+ * Wraps an underlying [[Stream]] with a function `f` that is called on each
+ * frame in the stream. The sequence of [[Frame]]s yielded by `f` will be
+ * inserted into the stream in order at the current position.
+ *
+ * @note This is essentially the same as the
+ *       [[scala.collection.TraversableLike.flatMap() flatMap]] function on
+ *       Scala collections, but with the limitation that the `flatMap`ped
+ *       function may not change the type of the mapped element (since H2
+ *       [[Stream]]s always consist of [[Frame]]s). To be properly monadic,
+ *       we would have the `flatMap`ped function return a new [[Stream]]
+ *       which would be spliced into this stream, but constructing streams
+ *       is much more complex (you have to make an
+ *       [[com.twitter.concurrent.AsyncQueue AsyncQueue]] and offer frames
+ *       to it one by one...), so we cheat a little by having the
+ *       `flatMap`pped function yield a [[Seq]] of frames instead.
+ * @note that in order to avoid violating flow control, `f` must either
+ *       take ownership over the frame and release it, or return it in
+ *       the returned sequence of frames.
+ * @param underlying the [[Stream]] wrapped by this proxy.
+ * @param f function called on each [[Frame]] in the underlying [[Stream]].
+ *
+ */
+class StreamFlatMap(underlying: Stream, f: Frame => Seq[Frame])
+  extends StreamProxy(underlying) {
+  private[this] var q = new mutable.Queue[Frame]()
+
+  /**
+   * @note that this stream is only empty if both the underlying stream
```

Comment:
Hmm...I decided to override `isEmpty` because it seems to me that a `Stream` probably shouldn't claim to be empty if `read()`ing from it will yield more frames? If `q` is non-empty, `read()` on this stream will yield frames, even if the `underlying` stream `isEmpty`. Which seems...surprising, if not actually incorrect?